File: java/src/main/java/com/penapereira/patterns/observer/Subject.java
package com.penapereira.patterns.observer;

import java.util.ArrayList;
import java.util.List;

/** Knows its observers and notifies them in attachment order. */
public class Subject {

    private final List<Observer> observers = new ArrayList<>();

    public void attach(Observer observer) {
        observers.add(observer);
    }

    public void detach(Observer observer) {
        observers.remove(observer);
    }

    protected void notifyObservers(int oldState, int newState) {
        for (Observer observer : List.copyOf(observers)) {
            observer.update(oldState, newState);
        }
    }
}

File: python/src/patterns/observer/subject.py
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .observer import Observer


class Subject:
    """Knows its observers and notifies them in attachment order."""

    def __init__(self) -> None:
        self._observers: list[Observer] = []

    def attach(self, observer: Observer) -> None:
        self._observers.append(observer)

    def detach(self, observer: Observer) -> None:
        self._observers.remove(observer)

    def notify_observers(self, old_state: int, new_state: int) -> None:
        for observer in list(self._observers):
            observer.update(old_state, new_state)
